launch_testing/launch_testing/__init__.py

```diff
@@ -17,52 +17,152 @@
 from launch.actions import EmitEvent
 from launch.actions import ExecuteProcess
 from launch.actions import RegisterEventHandler
+from launch.event_handlers import OnExecutionComplete
 from launch.event_handlers import OnProcessExit
+from launch.event_handlers import OnProcessIO
 from launch.events import Shutdown
 
-# for backward compatibility
-from launch_testing.legacy import create_handler  # noqa: F401
-from launch_testing.legacy import get_default_filtered_patterns  # noqa: F401
-from launch_testing.legacy import get_default_filtered_prefixes  # noqa: F401
-from launch_testing.legacy import get_rmw_output_filter  # noqa: F401
-from launch_testing.legacy import InMemoryHandler  # noqa: F401
-from launch_testing.legacy import UnmatchedOutputError  # noqa: F401
+from .output import create_output_check
 
 
 class LaunchTestService():
 
     def __init__(self):
-        self.__test_processes = []
-        self.__test_returncodes = OrderedDict()
+        self.__tests = OrderedDict()
+        self.__processes_rc = OrderedDict()
+
+    def _arm(self, test_name):
+        assert test_name not in self.__tests
+        self.__tests[test_name] = 'armed'
+
+    def _fail(self, test_name, reason):
+        self.__tests[test_name] = 'failed'
+        for test_name in self.__tests:
+            if self.__tests[test_name] == 'armed':
+                self.__tests[test_name] = 'dropped'
+        return EmitEvent(event=Shutdown(reason=reason))
+
+    def _succeed(self, test_name):
+        self.__tests[test_name] = 'succeeded'
+        if all(status == 'succeeded' for status in self.__tests.values()):
+            return EmitEvent(event=Shutdown(reason='all tests finished'))
+
+    def add_fixture_action(self, launch_description, action, required=False):
```

Comment:
Passing the launch_description in here was a bit off-putting to me at first, but I see the reason for it now, though I wish it were more transparent what was happening. If you look just at the calling code (like at one of the converted tests), you don't get a since as to what is being done to the launch description, you can assume it's adding the fixture/test/output_test or whatever, but I hadn't imagined it would add more actions until reading the code for it here.

A possible alternative would be to make this function simply generate and return a list of actions to be added, rather than adding them directly. Though I'm not sure if the launch description is required within this function in any other cases. For example, it might look like:

```
my_test_exec_action = ExecuteProcess(...)
fixture_action = lts.create_fixture_action(my_test_exec_action)
ld.add_action(fixture_action)
```

In that example `fixture_action` might be just the original action or a `GroupAction` that consists of several things.

I'm not 100% sold on my approach either, as I wouldn't be able to know without an example or very good docs if I should add the original action or just the result of `create_fixture_action()` instead.

Reply:
> Passing the launch_description in here was a bit off-putting to me at first...

It sort of is to me as well, but as I said before, I was trying to minimize the amount of code changes (because it's not at all unlikely we end up ditching this in the near future) and thus I kept the API they way it was. 

> A possible alternative would be to make this function simply generate and return a list of actions to be added, rather than adding them directly.

That is an option, sure. Another easy one (easy to use I mean) would be to have a `LaunchTestDescription` along with the `LaunchTestService` and make them wrap the `LaunchDescription` and `LaunchService` respectively. Not sure if it makes sense to spend time polishing this API though.